Fix decryption issue when postponing S/MIME encrypted mails

rayfordshire opened the pull request:
## What does this PR do?

I don't claim that I fully understand what the code does, but by symmetry between S/MIME end PGP this should be as presented here (and it fixes #3711 too!)

## What I (believe to) have found out:

The issue is that when postponing a Mail marked for S/MIME encryption and then recalling that encrypted message, then the headers seem to get corrupted.  Actually, I believe now that the headers are not corrupted they are the headers belonging to the outer message + some random(?) modification to set the MIME type to text/plain:

https://github.com/neomutt/neomutt/blob/01733225d4eed2a12b8400d37df881b3d94143c5/postpone/postpone.c#L466-L467

(Quest for the ambitious: Figure out why this is there and if it can be removed)

I believe what actually happened was, that the S/MIME message was *not decrypted at all*! A decryption code path existed for PGP but was simply missing for S/MIME.  One of the commits adds the same code path as for PGP but uses S/MIME.
I guess by sheer coincidence a later level decrypted the message (but that level did not care about any headers...)

## What else

The `dump_envelope()` crashed if given a null pointer, now fixed.  Also changed (IMHO enhanced) the debug methods to print if they got a null pointer. Take it or leave it I don't care.

In the code path the protected headers were also treated differently between PGP and S/MIME.  I don't understand what the code does but I'm a firm believer that there should be no difference between PGP and S/MIME.  (Hopefully this does not destroy some hidden invariance somewhere else...)

And a little treat to reduce duplication (and hopefully not forget one of PGP / S/MIME) the next time something gets updated.

## Conclusion

This fixes the postponing issue (at least I can no longer replicate it).

An honourable mention: during the debugging I noticed the following code:

https://github.com/neomutt/neomutt/blob/01733225d4eed2a12b8400d37df881b3d94143c5/ncrypt/crypt_gpgme.c#L2759-L2819

Here `tattach` contains the decrypted part including Content-Type etc headers.  But only some (e.g. `mime_headers`) are actually preserved, the rest is thrown away. (Incidentally, that was the layer which decrypted the postponed message later and one reason for the charset mismatch). Also fascinating(?) `decrypt_part()` passes its return value to its caller via a file descriptor...

And lastly, **kudos to all the maintainers**.

## What are the relevant issue numbers?

#3711

gahr commented on debug/email.c:
```diff
@@ -139,13 +145,17 @@ void dump_envelope(const struct Envelope *env)
+  if (!e)
+  {
+    mutt_debug(LL_DEBUG1, "\tNULL pointer\n");
+    return;
```
This leaks `buf`

gahr commented on debug/email.c:
```diff
@@ -242,14 +261,17 @@ void dump_param_list(const struct ParameterList *pl)
   mutt_debug(LL_DEBUG1, "Body\n");
 
+  if (!body)
+  {
```
This leaks `buf`

gahr commented on postpone/postpone.c:
```diff
@@ -418,13 +417,19 @@ static int create_tmp_files_for_attachments(FILE *fp_body, struct Buffer *file,
+        sec_type = mutt_is_application_pgp(b);
+      if (((WithCrypto & APPLICATION_SMIME) != 0) && !sec_type)
+        sec_type = mutt_is_application_smime(b);
+      if (sec_type & (SEC_ENCRYPT | SEC_SIGN))
```
I’m not convinced by the use of `sec_type` as a bool, using the fact that `SEC_NO_FLAGS` is 0. Wouldn’t it be better to test explicitely?

gahr commented:
Thanks, I’ll have a deeper look in the morning (on mobile now) - but it looks good from a first sight! A couple of minor things are commented inline.

rayfordshire commented:
Fixed the `buf` leaks, thanks for finding them.

>```c
>      SecurityFlags sec_type = SEC_NO_FLAGS;
>      if (((WithCrypto & APPLICATION_PGP) != 0) && !sec_type)
>        sec_type = mutt_is_application_pgp(b);
>      if (((WithCrypto & APPLICATION_SMIME) != 0) && !sec_type)
>        sec_type = mutt_is_application_smime(b);
>      if (sec_type & (SEC_ENCRYPT | SEC_SIGN))
>      { //...
>```
>
> I’m not convinced by the use of sec_type as a bool, using the fact that SEC_NO_FLAGS is 0. Wouldn’t it be better to test explicitely?

I'm also not a fan of it but the code I rewrote did that, so at least I was sure not to break anything (I still don't know what assumptions can be made about the flag/return value of `mutt_is_application_smime()` et al). If there is a better solution I'm happy to hear.

flatcap commented:
Rebased; no code changes

gahr commented:
> I'm also not a fan of it but the code I rewrote did that

Fair enough - I'd test against `SEC_NO_FLAGS` explicitely.

rayfordshire commented:
>> I'm also not a fan of it but the code I rewrote did that
>
>Fair enough - I'd test against SEC_NO_FLAGS explicitely.

Ok, I replaced `!sec_flag` with `sec_fag == SEC_NO_FLAGS`.
